src/priority_enumerator.jl

```diff
@@ -2,28 +2,58 @@
 Enumerates a context-free grammar up to a given depth in a breadth-first order.
 This means that smaller programs are returned before larger programs.
 """
-mutable struct ContextFreeBFSEnumerator <: ExpressionIterator 
+mutable struct ContextFreePriorityEnumerator <: ExpressionIterator 
     grammar::ContextFreeGrammar
     max_depth::Int
+    # Assigns a priority to a (partial or complete) tree.
+    priority_function::Function
+    # Expands a partial tree.
+    expand_function::Function
     sym::Symbol
+end 
+
+bfs_priority_function(tree::RuleNode, parent_value::Number) = parent_value + 1
+bfs_expand_heuristic(rules) = rules
+
+
+"""
+Returns a breadth-first search enumerator. Returns trees in the grammar in increasing order of size. 
+"""
+function get_bfs_enumerator(grammar::ContextFreeGrammar, max_depth::Int, sym::Symbol)::ContextFreePriorityEnumerator
+    expand_function(node, grammar, max_depth) = _expand(node, grammar, max_depth, bfs_expand_heuristic)
+    return ContextFreePriorityEnumerator(grammar, max_depth, bfs_priority_function, expand_function, sym)
+end
+
+dfs_priority_function(tree::RuleNode, parent_value::Number) = parent_value - 1
+dfs_expand_heuristic(rules) = rules
+
+
+"""
+Returns a depth-first search enumerator. Returns trees in the grammar in decreasing order of size.
+"""
+function get_dfs_enumerator(grammar::ContextFreeGrammar, max_depth::Int, sym::Symbol)::ContextFreePriorityEnumerator
+    expand_function(node, grammar, max_depth) = _expand(node, grammar, max_depth, dfs_expand_heuristic)
+    return ContextFreePriorityEnumerator(grammar, max_depth, dfs_priority_function, expand_function, sym)
 end
 
 
-function Base.iterate(iter::ContextFreeBFSEnumerator)
+function Base.iterate(iter::ContextFreePriorityEnumerator)
     # Priority queue with number of nodes in the program
-    pq :: PriorityQueue{RuleNode, Int} = PriorityQueue()
+    pq :: PriorityQueue{RuleNode, Any} = PriorityQueue()
```

Comment:
should this be `Any` or perhaps something like `Union{Int,Float}`? 
This could make a small performance difference that could accumulate

Reply:
Good point. Hmm. I see 3 options:

1. Keep the Int as a return. We can have tie-breaking with smart encodings of tuples. for example, if it should return a 2-tuple, we can turn it into `1000*[first number] + [second number]`, which would break ties if needed
2. Assume the fixed size of possible tuples and make it `Union{Int, Tuple{Int,Int}, Tuple{Int,Int,Int}}`.
3. Introduce a new struct `Result` that is flexible enough to handle all cases.

Maybe the first point gets us the most speed?
@jaapdejong15, could you encode your tuples into a scheme list above?

Reply:
@sebdumancic Tuples of arbitrary length can be defined using `Vararg`, e.g. Union{Real, Tuple{Vararg{Real}}}`, this is what I did in the latest commit. I also did some (limited) testing with int values and the different type signatures(Int/Any/current solution), but I couldn't find a significant performance difference between them.

Reply:
alright, let's then go with that option.

Reply:
Maybe then just use `Vararg{Real}` instead of `Union{Real, Tuple{Vararg{Real}}}`? Or does that complicates something?